pilot/pkg/config/mcp/controller.go

```diff
@@ -0,0 +1,212 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package coredatamodel
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"istio.io/fortio/log"
+	"istio.io/istio/pilot/pkg/model"
+	mcpclient "istio.io/istio/pkg/mcp/client"
+)
+
+var (
+	errNotFound    = errors.New("item not found")
+	errUnsupported = errors.New("this operation is not supported by mcp controller")
+)
+
+type CoreDataModel interface {
+	model.ConfigStoreCache
+	mcpclient.Updater
+}
+
+type Controller struct {
+	configStore              map[string]map[string]*sync.Map
+	eventCh                  chan func(model.Config, model.Event)
+	descriptorsByMessageName map[string]model.ProtoSchema
+}
+
+func NewController() CoreDataModel {
+	descriptorsByMessageName := make(map[string]model.ProtoSchema, len(model.IstioConfigTypes))
+	for _, config := range model.IstioConfigTypes {
+		descriptorsByMessageName[config.MessageName] = config
+	}
+
+	// Remove this when https://github.com/istio/istio/issues/7947 is done
+	configStore := make(map[string]map[string]*sync.Map)
+	for _, typ := range model.IstioConfigTypes.Types() {
+		configStore[typ] = make(map[string]*sync.Map)
+	}
+	return &Controller{
+		configStore:              configStore,
+		eventCh:                  make(chan func(model.Config, model.Event)),
+		descriptorsByMessageName: descriptorsByMessageName,
+	}
+}
+
+func (c *Controller) ConfigDescriptor() model.ConfigDescriptor {
+	return model.IstioConfigTypes
+}
+
+func (c *Controller) List(typ, namespace string) ([]model.Config, error) {
+	_, ok := c.ConfigDescriptor().GetByType(typ)
+	if !ok {
+		return nil, errors.New(fmt.Sprintf("List: unknown type %s", typ))
+	}
+	data, exists := c.configStore[typ]
+	if !exists {
+		log.Infof("List: config not found for the type %s", typ)
+		return nil, nil
+	}
+	out := make([]model.Config, 0, len(c.configStore[typ]))
+	if namespace == "" {
+		for _, ns := range data {
+			ns.Range(func(key, value interface{}) bool {
+				out = append(out, value.(model.Config))
+				return true
+			})
+		}
+	} else {
+		ns, exists := data[namespace]
+		if !exists {
+			return nil, nil
+		}
+		ns.Range(func(key, value interface{}) bool {
+			out = append(out, value.(model.Config))
+			return true
+		})
+	}
+	return out, nil
+}
+
+func (c *Controller) Apply(change *mcpclient.Change) error {
```

Comment:
MCP's `Apply()` delivers full state for the specified type. You'll need to erase all of the previous resources for the type before adding the new resources in the change list. Incrementally adding/removing resources without locking the entire store could lead to inconsistent reviews. Locking the entire store would allow for some reuse of the underlying map memory at the expense of holding locks longer.

Alternatively, you could build the new per-type state in a batch and swap it at the end (see below).

```golang
byType := make(map[string]map[string]model.Config) // by [namespace][name]

descriptor, ok := c.descriptorsByMessageName[change.MessageName] 
if !ok { /* return error */ }

schema, ok := c.ConfigDescriptor().GetByType(descriptor.Type)
if !ok { /* return error */ }

for _, obj := range change.Objects {
	// build model.Config from MCP resource
    // validate config.Config

    byNamespace, ok := byType[config.Namespace]
	if !ok {
		byNamespace = make(map[string]model.Config)
        byType[descriptor.Type][config.Namespace] = byNamespace
	} 
    byNamespace[config.Namespace][config.Name] = config   
}

c.configStoreMu.Lock()
c.configStore[descriptor.Type] = byNamespace
c.configStoreMu.Unlock
```


[Incremental MCP](https://github.com/envoyproxy/data-plane-api/blob/master/XDS_PROTOCOL.md#incremental-xds) and propagating per-resource versions is probably needed to avoid unnecessary memory allocations. We can add that later, possibly before MCP becomes the default.